driver-sync/src/main/com/mongodb/client/internal/ClientEncryptionImpl.java

```diff
@@ -171,6 +183,65 @@ public RewrapManyDataKeyResult rewrapManyDataKey(final Bson filter, final Rewrap
         return new RewrapManyDataKeyResult(collection.bulkWrite(updateModels));
     }
 
+    @Override
+    public BsonDocument createEncryptedCollection(final MongoDatabase database, final String collectionName,
+            final CreateCollectionOptions createCollectionOptions, final CreateEncryptedCollectionParams createEncryptedCollectionParams)
+            throws MongoUpdatedEncryptedFieldsException {
+        notNull("collectionName", collectionName);
+        notNull("createCollectionOptions", createCollectionOptions);
+        notNull("createEncryptedCollectionParams", createEncryptedCollectionParams);
+        MongoNamespace namespace = new MongoNamespace(database.getName(), collectionName);
+        Bson rawEncryptedFields = createCollectionOptions.getEncryptedFields();
+        if (rawEncryptedFields == null) {
+            throw new MongoConfigurationException(format("`encryptedFields` is not configured for the collection %s.", namespace));
+        }
+        CodecRegistry codecRegistry = options.getKeyVaultMongoClientSettings() == null
+                ? MongoClientSettings.getDefaultCodecRegistry()
+                : options.getKeyVaultMongoClientSettings().getCodecRegistry();
+        BsonDocument actualEncryptedFields = rawEncryptedFields.toBsonDocument(BsonDocument.class, codecRegistry).clone();
+        BsonValue fields = actualEncryptedFields.get("fields");
+        if (fields != null) {
+            if (!fields.isArray()) {
+                throw new MongoConfigurationException(format("`encryptedFields` is incorrectly configured for the collection %s."
+                        + " `encryptedFields.fields` must be an array, but is %s type.", namespace, fields.getBsonType()));
+            }
+            String kmsProvider = createEncryptedCollectionParams.getKmsProvider();
+            DataKeyOptions dataKeyOptions = new DataKeyOptions();
+            BsonDocument masterKey = createEncryptedCollectionParams.getMasterKey();
+            if (masterKey != null) {
+                dataKeyOptions.masterKey(masterKey);
+            }
+            String keyIdBsonKey = "keyId";
+            // any mutable non-thread-safe Boolean should do
+            AtomicBoolean dataKeyMayBeCreated = new AtomicBoolean();
```

Comment:
This method is required to ignore non-documents: [_"If `F` is not a document element, skip it."_](https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#create-encrypted-collection-helper:~:text=If%20F%20is%20not%20a%20document%20element%2C%20skip%20it.).

This follow the general approach: avoid validating in the driver what can be validated by the server (if a type system prevents specifying an invalid value via the driver API - that's fine).

The reason `fields.isArray()` is checked instead of relaying the check to the server is described in DRIVERS-2540 (I mentioned this in https://github.com/mongodb/mongo-java-driver/pull/1079#discussion_r1096460072).

Reply:
Thanks :+1: